### ISSUE TITLE
Specifying an additional thunk argument

### ISSUE THREAD

**Issue by iansinnott** [faceyspacey/redux-first-router#54]
**TL:DR:** Would be cool if I could inject a third argument into thunks, like this:

```js
const userThunk = async (dispatch, getState, api) => { // Notice the third argument
  const { slug } = getState().location.payload
  const data = await api.fetch(`/api/user/${slug}`) // Using the third argument
  const action = { type: 'USER_FOUND', payload: { user } }
  
  dispatch(action)
}

const routesMap = {
  USER: { path: '/user/:slug', thunk: userThunk  },
}

const api = {
  fetch(url) { /* Some custom logic... */ }
}

const { reducer, middleware, enhancer } = connectRoutes(history, routesMap, {
  extraThunkArgument: api
})

// etc...
```

## Why

* Consistency with redux-thunk. redux-thunk let's you [inject a third argument](https://github.com/gaearon/redux-thunk#injecting-a-custom-argument)
* Improved thunk flexibility (this is also the reason redux-thunk supports it)

By improved flexibility I mean you get to do things like:

* Mock API calls in tests (great for testing thunks)
* Share singletons across thunks without importing them
* Do other cool stuff I haven't though of...

For a quick example, let's say you are using Apollo Client without the react layer and you want to share the instance among thunks:

```js
import ApolloClient, { createNetworkInterface } from 'apollo-client'
const networkInterface = createNetworkInterface({ uri: 'https://example.com/graphql' })
const client = new ApolloClient({ networkInterface })

// Set up thunk middleware
import thunk from 'redux-thunk';
const thunkMiddleware = thunk.withExtraArgument({ client });

// Set everything else up and create the redux store... 
// (you can imagine what this would look like)
```

So I'm proposing we allow the special RFR thunks to support the same functionality:

```js
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap, {
  extraThunkArgument: { client }
})
```

## Conclusion

Let's do it? I'd be happy to PR this, but @faceyspacey I wanted to get your thoughts first.

**Comment by iansinnott** [comment]
OK, so I decided to just make a PR since I added this functionality locally 😊. And a PR is worth 1000 words right?

**Comment by faceyspacey** [comment]
...continued in the PR..

**Comment by faceyspacey** [comment]
This is now done:

https://gist.github.com/faceyspacey/2771c05a62a338fa6d3d23d76e6e7c5a

@iansinnott I left you some additional notes in your PR.
